### PR TITLE
Fix choose directive type inference

### LINKED ISSUE #4220: [lit-html] Choose directive doesn't check types correctly
### Which package(s) are affected?

Lit Core (lit / lit-html / lit-element / reactive-element)

### Description

When I use `choose` directive like this, I would expect typescript to throw error since `'test'` and `'random'` can't be assigned to `CheckoutStep` type 

```ts
type CheckoutStep = 'register' | 'delivery' | 'payment';

function render(step: CheckoutStep) {
    return choose(step, [
        ['test', () => 1],
        ['random', () => 2],
        ['register', () => 3],
    ]);
}
```

### Reproduction

There should be a type error but there isn't: [Typescript gist](https://lit.dev/playground/#project=W3sibmFtZSI6InJlcHJvLnRzIiwiY29udGVudCI6ImltcG9ydCB7aHRtbCwgY3NzLCBMaXRFbGVtZW50fSBmcm9tICdsaXQnO1xuaW1wb3J0IHtjdXN0b21FbGVtZW50LCBwcm9wZXJ0eX0gZnJvbSAnbGl0L2RlY29yYXRvcnMuanMnO1xuaW1wb3J0IHtjaG9vc2V9IGZyb20gJ2xpdC1odG1sL2RpcmVjdGl2ZXMvY2hvb3NlLmpzJztcblxudHlwZSBDaGVja291dFN0ZXAgPSAncmVnaXN0ZXInIHwgJ2RlbGl2ZXJ5JyB8ICdwYXltZW50JztcblxuQGN1c3RvbUVsZW1lbnQoJ3gtcmVwcm8nKVxuZXhwb3J0IGNsYXNzIFhSZXBybyBleHRlbmRzIExpdEVsZW1lbnQgeyAgXG4gIHJlbmRlclN0ZXAoc3RlcDogQ2hlY2tvdXRTdGVwKSB7XG4gICAgcmV0dXJuIGNob29zZShzdGVwLCBbXG4gICAgICAgIFsndGVzdCcsICgpID0-IDFdLFxuICAgICAgICBbJ3JhbmRvbScsICgpID0-IDJdLFxuICAgICAgICBbJ3JlZ2lzdGVyJywgKCkgPT4gM10sXG4gICAgXSk7XG4gIH1cblxuICByZW5kZXIoKSB7XG4gICAgcmV0dXJuIHRoaXMucmVuZGVyU3RlcCgncmVnaXN0ZXInKTtcbiAgfVxufVxuIn0seyJuYW1lIjoiaW5kZXguaHRtbCIsImNvbnRlbnQiOiI8IURPQ1RZUEUgaHRtbD5cbjxoZWFkPlxuICA8c2NyaXB0IHR5cGU9XCJtb2R1bGVcIiBzcmM9XCIuL3JlcHJvLmpzXCI-PC9zY3JpcHQ-XG48L2hlYWQ-XG48Ym9keT5cbiAgPHgtcmVwcm8-PC94LXJlcHJvPlxuPC9ib2R5PlxuIn0seyJuYW1lIjoicGFja2FnZS5qc29uIiwiY29udGVudCI6IntcbiAgXCJkZXBlbmRlbmNpZXNcIjoge1xuICAgIFwibGl0XCI6IFwiXjIuMC4wXCIsXG4gICAgXCJAbGl0L3JlYWN0aXZlLWVsZW1lbnRcIjogXCJeMS4wLjBcIixcbiAgICBcImxpdC1lbGVtZW50XCI6IFwiXjMuMC4wXCIsXG4gICAgXCJsaXQtaHRtbFwiOiBcIl4yLjAuMFwiXG4gIH1cbn0iLCJoaWRkZW4iOnRydWV9XQ)

### Workaround

It's seems like it's because `T` is inferred as `string` since it's used for both `value` and `cases` params
```ts
export declare const choose: <T, V>(value: T, cases: [T, () => V][]) => V;
```

It works as expected when I update the signature like this
```ts
export declare const choose: <T, V, K extends T>(value: T, cases: [K, () => V][]) => V;
```

### Is this a regression?

No or unsure. This never worked, or I haven't tried before.

### Affected versions

This never worked

### Browser/OS/Node environment

OS: Windows 10
Node version: 18.12.1
npm version: 8.19.2

### PR DESCRIPTION
**Pull request by remziatay**
Fixes #4220

### CODE REVIEW

File: packages/lit-html/src/directives/choose.ts [hunk 1]
```diff
@@ -29,9 +29,9 @@
  * }
  * ```
  */
-export const choose = <T, V>(
+export const choose = <T, V, K extends T = T>(
```
Comment (rictic) [inline packages/lit-html/src/directives/choose.ts:32]: Nice! Yeah this fixes the issue. [Playground demo](https://lit.dev/playground/#project=W3sibmFtZSI6InJlcHJvLnRzIiwiY29udGVudCI6ImltcG9ydCB7aHRtbCwgY3NzLCBMaXRFbGVtZW50fSBmcm9tICdsaXQnO1xuaW1wb3J0IHtjdXN0b21FbGVtZW50LCBwcm9wZXJ0eX0gZnJvbSAnbGl0L2RlY29yYXRvcnMuanMnO1xuZGVjbGFyZSBjb25zdCBjaG9vc2U6IDxULCBWLCBLIGV4dGVuZHMgVCA9IFQ-KFxuICB2YWx1ZTogVCxcbiAgY2FzZXM6IEFycmF5PFtLLCAoKSA9PiBWXT4sXG4gIGRlZmF1bHRDYXNlPzogKCkgPT4gVlxuKSA9PiB1bmtub3duO1xuXG50eXBlIENoZWNrb3V0U3RlcCA9ICdyZWdpc3RlcicgfCAnZGVsaXZlcnknIHwgJ3BheW1lbnQnO1xuXG5AY3VzdG9tRWxlbWVudCgneC1yZXBybycpXG5leHBvcnQgY2xhc3MgWFJlcHJvIGV4dGVuZHMgTGl0RWxlbWVudCB7ICBcbiAgcmVuZGVyU3RlcChzdGVwOiBDaGVja291dFN0ZXApIHtcbiAgICByZXR1cm4gY2hvb3NlKHN0ZXAsIFtcbiAgICAgICAgWyd0ZXN0JywgKCkgPT4gMV0sXG4gICAgICAgIFsncmFuZG9tJywgKCkgPT4gMl0sXG4gICAgICAgIFsncmVnaXN0ZXInLCAoKSA9PiAzXSxcbiAgICBdKTtcbiAgfVxuXG4gIHJlbmRlcigpIHtcbiAgICByZXR1cm4gdGhpcy5yZW5kZXJTdGVwKCdyZWdpc3RlcicpO1xuICB9XG59XG4ifSx7Im5hbWUiOiJpbmRleC5odG1sIiwiY29udGVudCI6IjwhRE9DVFlQRSBodG1sPlxuPGhlYWQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vcmVwcm8uanNcIj48L3NjcmlwdD5cbjwvaGVhZD5cbjxib2R5PlxuICA8eC1yZXBybz48L3gtcmVwcm8-XG48L2JvZHk-XG4ifSx7Im5hbWUiOiJwYWNrYWdlLmpzb24iLCJjb250ZW50Ijoie1xuICBcImRlcGVuZGVuY2llc1wiOiB7XG4gICAgXCJsaXRcIjogXCJeMi4wLjBcIixcbiAgICBcIkBsaXQvcmVhY3RpdmUtZWxlbWVudFwiOiBcIl4xLjAuMFwiLFxuICAgIFwibGl0LWVsZW1lbnRcIjogXCJeMy4wLjBcIixcbiAgICBcImxpdC1odG1sXCI6IFwiXjIuMC4wXCJcbiAgfVxufSIsImhpZGRlbiI6dHJ1ZX1d)

**Review by rictic — CHANGES_REQUESTED** [review]
Mind adding a regression test to the `choose_test.ts`? Something like:

```ts
  if (false as boolean) {
    // Type-only regression test of https://github.com/lit/lit/issues/4220
    type CheckoutStep = 'register' | 'delivery' | 'payment';
    let step = 'register' as CheckoutStep;
    return choose(step, [
      // @ts-expect-error 'test' is not assignable to 'CheckoutStep'
      ['test', () => 1],
      // @ts-expect-error 'random' is not assignable to 'CheckoutStep'
      ['random', () => 2],
      // this should compile fine
      ['register', () => 3],
    ]);
  }
```

**Comment by rictic** [comment]
Also please add a changeset by running `npm run changeset` at the root of the monorepo. This seems like a patch level change to lit-html, with a description like: `Improved the type of the choose directive.`

**Comment by remziatay** [comment]
> Mind adding a regression test to the `choose_test.ts`? Something like:

Of course! Done

**Review by augustjk — APPROVED** [review]
Thank you for the PR and sorry it took so long to get it in! I fixed the build error for the test and made the changelog a bit more detailed.

File: packages/lit-html/src/test/directives/choose_test.ts [hunk 2]
```diff
@@ -46,8 +46,8 @@ suite('choose', () => {
     );
   });
 
-  if (false as boolean) {
-    // Type-only regression test of https://github.com/lit/lit/issues/4220
+  // Type-only regression test of https://github.com/lit/lit/issues/4220
+  test.skip('type-only: correctly infers typ eof possible cases from value', () => {
```
Comment (AndrewJakubowicz) [inline packages/lit-html/src/test/directives/choose_test.ts:50]: "typ eof" -> "type of"

Comment (augustjk) [reply packages/lit-html/src/test/directives/choose_test.ts:50]: oops! thank you!

File: .changeset/good-grapes-double.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'lit-html': patch
+---
+
+Improved the type inferece of the `choose()` directive to properly restrict the case type inferred from provided value.
```
Comment (augustjk) [inline .changeset/good-grapes-double.md:5]: Adding some additional info for potential user breakage as suggested by @AndrewJakubowicz 
```suggestion
Improved the type inferece of the `choose()` directive to properly restrict the case type inferred from provided value. **Note**: If this change creates a type error in your code, there must have been an unreachable case that can be removed, or the type of your `value` might be missing a valid case in the union.
```

Comment (AndrewJakubowicz) [reply .changeset/good-grapes-double.md:5]: Beautiful!

**Review by AndrewJakubowicz — APPROVED** [review]
Great! Thank you!

**Pull request merged**